include/itkVkMultiResolutionPyramidImageFilter.hxx

```diff
@@ -0,0 +1,210 @@
+/*=========================================================================
+ *
+ *  Copyright NumFOCUS
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *         http://www.apache.org/licenses/LICENSE-2.0.txt
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ *
+ *=========================================================================*/
+#ifndef itkVkMultiResolutionPyramidImageFilter_hxx
+#define itkVkMultiResolutionPyramidImageFilter_hxx
+
+#include "itkCastImageFilter.h"
+#include "itkDiscreteGaussianImageFilter.h"
+#include "itkGaussianOperator.h"
+#include "itkMacro.h"
+#include "itkResampleImageFilter.h"
+#include "itkShrinkImageFilter.h"
+#include "itkIdentityTransform.h"
+
+#include "itkMath.h"
+
+namespace itk
+{
+template <typename TInputImage, typename TOutputImage>
+void
+VkMultiResolutionPyramidImageFilter<TInputImage, TOutputImage>::GenerateData()
+{
+  // Mostly reimplements MultiResolutionPyramidImageFilter::GenerateData
+
+  // Get the input and output pointers
+  InputImageConstPointer inputPtr = this->GetInput();
+
+  // Create caster, smoother and resampleShrinker filters
+  using CasterType = CastImageFilter<TInputImage, TOutputImage>;
+
+  using ImageToImageType = ImageToImageFilter<TOutputImage, TOutputImage>;
+  using ResampleShrinkerType = ResampleImageFilter<TOutputImage, TOutputImage>;
+  using ShrinkerType = ShrinkImageFilter<TOutputImage, TOutputImage>;
+
+  auto                               caster = CasterType::New();
+  auto                               spatialSmoother = SpatialSmootherType::New();
+  auto                               fftSmoother = FFTSmootherType::New();
```

Comment:
For another pull request: perhaps it would make sense to have these workhorse smoothers (`spatialSmoother` and `fftSmoother`) smart pointers be members of the present object, so that if the present object is re-used, so would they be.  In that case, the check whether one or the other is actually used in the first use of the present object would be less important; we might `New` them both as part of the constructor (or first-use) of the present object and keep them around.

Reply:
I've gone ahead and added this as a small change in the existing PR so that both smoothers are constructed as members of the pyramid filter at the time of its construction. Note that we reasonably expect developers using ITKVkFFTBackend are dealing with large images already, so I do not think it is worthwhile to spend additional time deciding how to selectively initialize the smoothers as the performance benefit is almost certain to be negligible in comparison.